api/app.py

```diff
@@ -20,37 +20,37 @@
 from core.util import LanguageCodes
 from flask.ext.babel import Babel
 
-app = Flask(__name__)
-
-testing = 'TESTING' in os.environ
-db_url = Configuration.database_url(testing)
-# Initialize a new database session unless we were told not to
-# (e.g. because a script already initialized it).
-autoinitialize = os.environ.get('AUTOINITIALIZE') != 'False'
-if autoinitialize:
-    SessionManager.initialize(db_url)
-session_factory = SessionManager.sessionmaker(db_url)
-_db = flask_scoped_session(session_factory, app)
-if autoinitialize:
-    SessionManager.initialize_data(_db)
 
+app = Flask(__name__)
+app._db = None
 app.config['BABEL_DEFAULT_LOCALE'] = LanguageCodes.three_to_two[Configuration.localization_languages()[0]]
 app.config['BABEL_TRANSLATION_DIRECTORIES'] = "../translations"
 babel = Babel(app)
 
-import routes
-import admin.routes
+@app.before_first_request
+def initialize_database(autoinitialize=True):
+    testing = 'TESTING' in os.environ
+    db_url = Configuration.database_url(testing)
+    if autoinitialize:
+        SessionManager.initialize(db_url)
+    session_factory = SessionManager.sessionmaker(db_url)
+    _db = flask_scoped_session(session_factory, app)
+    app._db = _db
+    if autoinitialize:
+        SessionManager.initialize_data(_db)
 
-log_level = LogConfiguration.initialize(_db, testing=testing)
-debug = log_level == 'DEBUG'
+    log_level = LogConfiguration.initialize(_db, testing=testing)
+    debug = log_level == 'DEBUG'
+    app.config['DEBUG'] = True
+    app.debug = debug
+    _db.commit()
+    logging.getLogger().info("Application debug mode==%r" % app.debug)
 
-logging.getLogger().info("Application debug mode==%r" % debug)
-app.config['DEBUG'] = debug
-app.debug = debug
+import routes
+import admin.routes
 
 def run(url=None):
-    base_url = ConfigurationSetting.sitewide(_db, Configuration.BASE_URL_KEY)
-    base_url = url or base_url.value or u'http://localhost:6500/'
+    base_url = url or u'http://localhost:6500/'
```

Comment:
We can no longer use the database BASE_URL_KEY as the URL, because that would require initializing a database connection before calling app.run(), which is what causes the problem. This could be a big problem or it could be nothing; I'm not sure.